### PR TITLE
Rules to add achievements for bonus categories

### LINKED ISSUE #697: Rules to add achievements for bonus categories
Create new model

* Rules (to keep the rule, ex. add an achievement for each 2 bonus points of category)
* When adding a bonus, check that rules and add an achievement if necessary

### PR DESCRIPTION
**Pull request by nyet-ty**
Resolve #697 

## PR includes

-   [x] Feature

### CODE REVIEW

File: prisma/migrations/20240424150211_bonus_rule_for_crew_achievement/migration.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+-- AlterTable
+ALTER TABLE "Achievement" ADD COLUMN     "bonusRuleId" TEXT;
+
+-- CreateTable
+CREATE TABLE "BonusRule" (
+    "id" TEXT NOT NULL,
+    "bonuceAmountForAchievement" INTEGER NOT NULL,
```
Comment (9teen90nine) [inline prisma/migrations/20240424150211_bonus_rule_for_crew_achievement/migration.sql:7]: bonus?

File: prisma/migrations/20240424150211_bonus_rule_for_crew_achievement/migration.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    "cathegoryId" TEXT NOT NULL,
```
Comment (9teen90nine) [inline prisma/migrations/20240424150211_bonus_rule_for_crew_achievement/migration.sql:8]: category?

**Review by 9teen90nine — CHANGES_REQUESTED** [review]

File: prisma/schema.prisma [hunk 3]
```diff
@@ -237,6 +237,13 @@ model Role {
   updatedAt DateTime @default(dbgenerated("timezone('utc'::text, now())")) @updatedAt
 }
 
+model BonusRule {
+  id                         String        @id @default(nanoid())
```
Comment (tsumo) [inline prisma/schema.prisma:241]: formatting is broken

File: prisma/schema.prisma [hunk 4]
```diff
@@ -237,6 +237,13 @@ model Role {
...
+  bonusAmountForAchievement Int
+  categoryId                String
+  crewAchievement            Achievement[]
```
Comment (tsumo) [inline prisma/schema.prisma:244]: just achievement here

File: src/modules/bonusPointsMethods.ts [hunk 5]
```diff
@@ -5,6 +5,7 @@ import { config } from '../config';
 
 import { ChangeBonusPoints, GetAchievements } from './bonusPointsSchemas';
 import { BonusPointsAchievement } from './bonusPointsTypes';
+import { achievementMethods } from './achievementMethods';
 
 export const bonusPointsMethods = {
     change: async (data: ChangeBonusPoints, sessionUserId: string): Promise<User> => {
```
Comment (tsumo) [inline src/modules/bonusPointsMethods.ts:11]: Needs a check that tech admin exists

**Review by tsumo — CHANGES_REQUESTED** [review]

File: prisma/schema.prisma [hunk 6]
```diff
@@ -359,7 +366,9 @@ model Achievement {
 
   achievements UserAchievement[]
 
-  createdAt DateTime @default(dbgenerated("timezone('utc'::text, now())")) @db.Timestamp()
+  createdAt   DateTime   @default(dbgenerated("timezone('utc'::text, now())")) @db.Timestamp()
```
Comment (tsumo) [inline prisma/schema.prisma:369]: Let's keep createdAt as the last line

File: src/modules/bonusPointsMethods.ts (hunk 5 above)
Comment (tsumo) [reply src/modules/bonusPointsMethods.ts:11]: Still need a check

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/modules/bonusPointsMethods.ts [hunk 7]
```diff
@@ -23,6 +25,46 @@ export const bonusPointsMethods = {
...
             }),
         ]);
+        if (data.achievementCategory && config.techAdminId) {
+            const techAdmin = await prisma.user.findUnique({ where: { id: config.techAdminId } });
+
+            if (!techAdmin) {
+                Sentry.captureException(`No tech admin with id ${config.techAdminId} from config`);
+                return user;
+            }
+            const bonusesInCathegory = await prisma.$queryRaw<Array<{ sum: number }>>`
```
Comment (tsumo) [inline src/modules/bonusPointsMethods.ts:35]: Category

**Review by tsumo — CHANGES_REQUESTED** [review]

**Review by tsumo — APPROVED** [review]

**Review by 9teen90nine — APPROVED** [review]

**Pull request merged**